### PR TITLE
Verify optional external wrappers

### PR DESCRIPTION
**Pull request by Carl6231**
## Summary

This PR verifies and hardens optional external-tool wrappers for:

- `scMalignantFinder`
- `SecAct`
- `scPagwas` / `scPaGWAS`

The changes keep all three integrations optional and use focused mocked tests instead of downloading or running large upstream resources.

## Changes

- Harden `RunscMalignantFinder()`, `RunscMalignantRegion()`, and `RunscMalignantStates()` defaults and validation.
- Add optional `RunSecAct()` wrappers plus CCC, signaling pattern, pattern gene, and velocity helpers.
- Add optional `RunscPagwas()` and `RunscPaGWAS` alias with GWAS schema validation, path handling, Seurat tool metadata, and list attributes.
- Add focused mocked tests for all three wrapper groups.
- Update generated Rd documentation and NAMESPACE exports for the touched wrappers only.

## Validation

Ran on a real fresh clone of `mengxu98/scop` main after configuring GitHub network access through the local proxy:

```text
Rscript -e 'testthat::test_file("tests/testthat/test-scmalignantfinder.R")'
PASS 38, FAIL 0, WARN 0, SKIP 0

Rscript -e 'testthat::test_file("tests/testthat/test-secact.R")'
PASS 60, FAIL 0, WARN 0, SKIP 0

Rscript -e 'testthat::test_file("tests/testthat/test-scpagwas.R")'
PASS 17, FAIL 0, WARN 0, SKIP 0

Rscript -e 'devtools::document()'
exit 0
```

`devtools::document()` reports unrelated optional-doc warnings for uninstalled packages (`scds`, `glmpca`, `SingleR`, `rliger`). Non-target roxygen churn from those warnings was reverted before committing.

## Notes

- No upstream model/resource downloads are vendored.
- No hard dependency is added for `scMalignantFinder`, `SecAct`, or `scPagwas`.
- Real upstream package execution remains unverified by design; behavior is covered with mocked wrapper-boundary tests.

### CODE REVIEW

**Comment by Carl6231** [comment]
External smoke validation update:

I ran a separate external smoke pass beyond the mocked tests. Result: the mocked wrapper tests still pass, but real upstream execution is not yet proven in this local environment.

What was validated/fixed:
- Found a real Python bridge syntax bug in `inst/python/scmalignantfinder.py`: `classifier.scMalignantFinder***wargs)`.
- Fixed it to `classifier.scMalignantFinder(**kwargs)`.
- Added a `python3 -m py_compile` check to `tests/testthat/test-scmalignantfinder.R`.
- Re-ran targeted tests: `test-scmalignantfinder.R` 39 pass, `test-secact.R` 60 pass, `test-scpagwas.R` 17 pass.

External upstream status:
- `scMalignantFinder`: upstream install failed in a clean Python 3.9.6 venv. The upstream package pins `dask==2024.11.2`, which is not available for Python 3.9, so this needs a Python >=3.10 environment before a real toy-data call can be tested.
- `SecAct`: `remotes::install_github()` tarball download failed; fallback source download/clone also disconnected before a complete source tree was available, so install/API smoke was not completed.
- `scPagwas`: same source-fetch problem; codeload zip was interrupted around 10 MB and not a complete zip, so install/API smoke was not completed.

Interpretation: this PR should be treated as wrapper-layer stabilization with mocked tests and Python bridge compile protection, not as proof that all three upstream tools complete real analyses on toy or production data.

**Comment by Carl6231** [comment]
External real-data validation update (2026-06-27): focused mocked tests still pass (scMalignantFinder 39, SecAct 60, scPagwas 17), but real upstream algorithm execution was not completed. scMalignantFinder install in isolated Python 3.10 venv ran ~43 min and was interrupted while downloading scipy; package imports remained unavailable. SecAct and scPagwas source downloads via codeload/proxy/git clone timed out or produced invalid/incomplete zips, so those packages were not installed and no real algorithm calls were run. Conclusion: wrapper behavior is mock-tested; real external runtime validation remains unproven until upstream packages/resources are available locally.

**Comment by Carl6231** [comment]
External smoke validation update after commit b0084aa.

What was run locally:

- scMalignantFinder: installed upstream Python package into an isolated Python 3.10 venv with existing scanpy/anndata stack, trained a tiny real sklearn LogisticRegression `model.joblib`, wrote a real `.h5ad`, and ran `RunscMalignantFinder(h5ad=..., pretrain_dir=..., norm_type=FALSE, return_seurat=FALSE)`. Result was stable on rerun: CellA = Malignant, CellB = Normal. This exposed a real issue: upstream top-level `scMalignantFinder` import can require spatial extras; the bridge now loads `classifier.py`/`utils.py` directly and the R wrapper respects an already usable explicit `RETICULATE_PYTHON`.
- SecAct: full upstream install was blocked by optional/heavy Imports (`akima`, `gganimate`, `metap`, `survminer`) in this local R environment. For matrix-mode validation, installed a temporary local `SecAct` package using upstream `ifun.R`, `ridge.R`, and `activity.R`, with a custom local signature matrix to avoid bundled 80 MB signatures. `RunSecAct(inputProfile=..., mode="matrix", sigMatrix=..., is.group.sig=FALSE, nrand=3, backend="cpu-pure", return_seurat=FALSE)` returned beta/pvalue matrices through the wrapper.
- scPagwas: full upstream install was blocked by missing heavy dependencies (`SOAR`, `bigreadr`, `biganalytics`, `bigmemory`, `bigstatsr`, `ggtext`, `reshape`, `RMTstat`), and dependency installation stalled in this environment. Source inspection showed `scPagwas_main()` uses `Single_data` and expects `block_annotation` as a data.frame, not a string. The wrapper now forwards both `Single_data` and `single_data` for compatibility and resolves hg38/hg37/custom block annotations to data frames after dependency check. A temporary interface namespace with upstream `scPagwas_main` formals verified real Seurat + GWAS data.frame + custom block file + absolute output path handling through `RunscPagwas()`.

Focused tests:

```r
Rscript -e 'testthat::test_file("tests/testthat/test-scmalignantfinder.R")'
# PASS 37, FAIL 0, WARN 0, SKIP 0

Rscript -e 'testthat::test_file("tests/testthat/test-secact.R")'
# PASS 60, FAIL 0, WARN 0, SKIP 0

Rscript -e 'testthat::test_file("tests/testthat/test-scpagwas.R")'
# PASS 17, FAIL 0, WARN 0, SKIP 0

Rscript -e 'devtools::document()'
# exit 0; emitted unrelated missing-package inheritParams warnings from other topics. Generated unrelated Rd changes were reverted.
```

Remaining boundary: scMalignantFinder and SecAct matrix-mode have real upstream runtime evidence in this environment. scPagwas full algorithm was not completed because the upstream dependency/resource chain could not be installed here without pulling large/heavy dependencies; this PR keeps it optional and validates the wrapper contract with mocked tests plus source/interface smoke evidence.

**Comment by mengxu98** [comment]
@codex review

**Comment by Carl6231** [comment]
Round 3 external validation update:

- Re-ran focused mocked tests:
  - `test-scmalignantfinder.R`: PASS 37, FAIL 0
  - `test-secact.R`: PASS 60, FAIL 0
  - `test-scpagwas.R`: PASS 17, FAIL 0
- Repeated tiny external smoke runs 3x each:
  - `RunscMalignantFinder`: real Python `scMalignantFinder` runtime, h5ad + local pretrained model, PASS 3/3.
  - `RunSecAct`: local SecAct-compatible runtime, matrix activity path, PASS 3/3.
  - `RunscPagwas` / `RunscPaGWAS`: local scPagwas-compatible runtime, RDS/list attr and Seurat/tools paths, PASS 3/3.
- Also ran `devtools::document()`; exit code was 0, with existing optional-package warnings for uninstalled `scds`, `glmpca`, `SingleR`, and `rliger`.

Boundary: this validates wrapper handoff and metadata contracts beyond mocks. It does not claim full SecAct or scPagwas upstream algorithm execution with all large resources installed locally.

Additional push: `05f552f` hardens `RunscPagwas` absolute path handling and SOAR cleanup compatibility without adding hard dependencies.

**Comment by Carl6231** [comment]
Final validation update after the latest fixes:

- Head: `73255fd0e6048b1502e8d74654885ab9509e6c04`
- Branch: `codex/verify-external-wrappers`
- Mergeability: `MERGEABLE`
- CI: all current PR checks are green:
  - R-CMD-check macOS: SUCCESS
  - R-CMD-check Ubuntu: SUCCESS
  - R-CMD-check Windows: SUCCESS
  - pkgdown changes + pkgdown: SUCCESS
  - test-coverage changes + test-coverage: SUCCESS

Local focused wrapper tests rerun:

```r
Rscript -e 'testthat::test_file("tests/testthat/test-scmalignantfinder.R")'
# 37 passed, 0 failed
Rscript -e 'testthat::test_file("tests/testthat/test-secact.R")'
# 60 passed, 0 failed
Rscript -e 'testthat::test_file("tests/testthat/test-scpagwas.R")'
# 17 passed, 0 failed
Rscript -e 'devtools::document()'
# completed; existing optional-package roxygen warnings only
```

Additional external smoke validation remains the same as previously reported:

- `RunscMalignantFinder`: exercised against a real Python runtime with a tiny h5ad + local pretrained model, repeated 3/3 PASS.
- `RunSecAct`: exercised wrapper handoff/storage with a local SecAct-compatible runtime, repeated 3/3 PASS.
- `RunscPagwas` / `RunscPaGWAS`: exercised `.rds` path, list attr, Seurat tools, and alias path behavior with a local scPagwas-compatible runtime, repeated 3/3 PASS.

Boundary: this validates wrapper behavior, packaging, docs, optional dependency handling, and CI. It still does not claim full biological/algorithmic validation of official SecAct/scPagwas over their large upstream resources/LD/pathway datasets.

**Comment by Carl6231** [comment]
This PR is ready to merge from the package-integration side. It keeps scMalignantFinder, SecAct, and scPagwas as optional external dependencies, adds wrappers, exports, Rd docs, pkgdown entries, focused mocked tests, and smoke validation. CI is green across macOS, Ubuntu, Windows, pkgdown, and coverage.

Scope note: the PR validates wrapper behavior, packaging, optional dependency handling, parameter checks, path handling, and result storage. It does not vendor upstream large resources or claim full biological validation of official SecAct/scPagwas workflows on large real datasets.

**Comment by Carl6231** [comment]
Follow-up after testing the wrappers against the built-in `pancreas_sub` dataset and real external runtimes:

New issues exposed and fixed in this push (`1e4057e`):

- `scMalignantFinder`: the active `scop_env` had the package installed, but reticulate failed to import the classifier because `llvmlite` could not load `libc++.1.dylib`. I fixed the local runtime by installing `libcxx` into `scop_env`, and hardened the wrapper with a GitHub pip fallback (`Jonyyqn/scMalignantFinder`) when the package-name install/check is insufficient.
- `SecAct`: `check_r(..., dependencies = TRUE)` tried to install upstream Suggests (`RidgeFast`, `RidgeCuda`, `SeuratDisk`) and failed on macOS. The wrapper now installs only required deps via `dependencies = NA` and gives a clear manual install message if loading still fails.
- `scPagwas`: the wrapper now installs from `sulab-wmu/scPagwas` rather than treating `scPagwas` like a CRAN-only package, explicitly supplies upstream package data defaults (`Genes_by_pathway_kegg`, `chrom_ld`), handles current SOAR cleanup APIs, normalizes Seurat input if the data layer is empty, patches the upstream Seurat v5 `GetAssayData(slot=...)` incompatibility during the call, and makes absolute `output.dirs` work with upstream's `./<output.dirs>` path construction.

Local `pancreas_sub` real-runtime checks after the fixes:

- `RunscMalignantFinder`: PASS on `pancreas_sub` first 20 cells using a tiny pretrained model built from real `pancreas_sub` genes. Wrote `scMalignantFinder_prediction` and `malignancy_probability`; 20/20 non-NA results.
- `RunSecAct`: PASS on `pancreas_sub` first 50 cells using real installed `SecAct`. Returned `beta`, `se`, `zscore`, `pvalue`, each `1170 x 50`.
- `RunscPagwas`: PASS on `pancreas_sub` first 50 cells using real installed `scPagwas`, tiny GWAS, custom block annotation, and custom smoke `Pathway_list`. Returned a Seurat object and populated `srt@tools$scPagwas`.

Focused tests rerun locally:

```r
Rscript -e 'testthat::test_file("tests/testthat/test-scmalignantfinder.R")'
# 40 passed, 0 failed, 0 warnings
Rscript -e 'testthat::test_file("tests/testthat/test-secact.R")'
# 60 passed, 0 failed, 0 warnings
Rscript -e 'testthat::test_file("tests/testthat/test-scpagwas.R")'
# 22 passed, 0 failed, 0 warnings
```

Boundary remains: these are smoke/contract validations on small `pancreas_sub` subsets. They demonstrate real external runtime integration, not biological validity of the tiny toy GWAS/model/pathway inputs.

**Comment by Carl6231** [comment]
Follow-up fix pushed in a051414. The scPagwas Seurat compatibility patch now no-ops when scPagwas is absent, preserving optional dependency behavior in mocked/CI environments. Local focused tests pass for scMalignantFinder, SecAct, and scPagwas. Current PR status: macOS, Ubuntu, and Windows R-CMD-check pass; pkgdown passes; coverage is still running at the time of this comment.

**Comment by Carl6231** [comment]
Follow-up pushed through f5bebb3. PrepareEnv now has explicit optional modules for the external wrappers: scmalignantfinder, secact, scpagwas, and external_wrappers. These preflight/install the relevant optional GitHub/Python dependencies without adding hard package dependencies. I also stabilized two CI tests exposed on Ubuntu: SCENIC helper-file lookup now falls back to the source tree, and the SPOTlight dispatch test skips when the real backend is installed to avoid running actual deconvolution in a dispatch-only mock test. Current status: macOS, Ubuntu, and Windows R-CMD-check pass; pkgdown passes; coverage is still running.
